src/Microsoft.Security.Utilities.Benchmarks/SecretMaskerDetectionBenchmarks.cs

```diff
@@ -5,42 +5,85 @@
+        // the EasyPrefix and HardPrefix cases. This is intended to make
+        // positive hits less concentrated in the profiling.
+        public const int SecretPrefixSize = 10 * 1024;
+
+        // Whether to use the high performance scanner. Normally, the high
+        // performance scanner is always used, but there's a test hook to turn
+        // it off, which we use to see how much it is optimizing.
+        [Params(true, false)]
+        public bool UseHighPerformanceScanner { get; set; } = true;
+
+        // Whether to generate correlating ids for each match. Setting this to
+        // true will contribute fixed hash production overhead to all the
+        // scanners. Uncomment 'Params' attribute to benchmark with and without
+        // it.
+        //
+        //[Params(true,  false)]
+        public bool GenerateCorrelatingIds { get; set; } = false;
+
+        // The regex engine to use. The DotNet engine is faster than RE2 on
+        // modern .NET, and RE2 is faster than DotNet on .NET Framework for
+        // large input. Uncomment Params attribute to benchmark both engines,
+        // Otherwise, we'll use the better default for the current platform.
+        //
+        //[Params(RegexEngine.DotNet, RegexEngine.RE2)]
+        public RegexEngine RegexEngine { get; set; } =
+#if NET
+            RegexEngine.DotNet;
+#else
+            RegexEngine.RE2;
+#endif
+
+        [ParamsAllValues]
+        public Case Case { get; set; } = Case.HardPrefix;
 
         // Use a fixed seed to ensure different runs use the same data.
-        private readonly Random rng = new Random(Seed: 42);
+        private readonly Random _rng = new Random(Seed: 42);
 
-        protected SecretMaskerDetectionBenchmarks()
-        {
-            RandomExamples = GenerateExamples(random: true);
-            NonRandomExamples = GenerateExamples(random: false);
-        }
+        protected List<string> Examples { get; private set; } = null!;
+        protected SecretMasker Masker { get; private set; } = null!;
+        protected abstract IEnumerable<RegexPattern> RegexPatterns { get; }
 
         private string GeneratePrefix(int size, bool random)
         {
             string prefix;
 
             if (random)
             {
-                // This is close to the worst case scenario. Every base64
-                // alphabet char is equally likely to appear in the input,
-                // making signature sniffing harder.
                 var data = new byte[size / 4 * 3]; // Account for base64 encoding overhead.
-                rng.NextBytes(data);
+                _rng.NextBytes(data);
                 prefix = Convert.ToBase64String(data);
             }
             else
             {
-                // This is close to the best case scenario. This char does not appear in
-                // any identifiable key signature, making signature sniffing easier.
                 prefix = new string('%', size);
```

Comment:
>'%' [](http://example.com/codeflow?start=36&length=3)

SEV4 maybe change this to a unicode character or something more guaranteed never to appear in a regex, for example `\t`. someday, perhaps a range of chars, which includes things like `\0` could be interesting (but then again maybe this data diversity isn't helpful in the foreseeable future)

Reply:
I'll leave this for now. I'm marginally worried that an exotic char will trigger atypical code paths. In general, I think there's a rich opportunity to get more sophisticated here and figure out how to synthesize data that represents real scenarios well.